### PR TITLE
petsc fix hdf5

### LINKED ISSUE #15553: petsc build error after update to hdf5 1.12
### Spack version
Current `develop` branch.

### Steps to reproduce the issue

```console
$ spack install petsc@3.7.7 ^superlu_dist@5.1.3_
```
(The `superlu_dist` version is required until #15551 is merged, an related issue.)

leads to:
```console
/gpfs/alpine/fus123/proj-shared/kaig1/spack/opt/spack/linux-rhel7-power9le/gcc-8.1.1/hdf5-1.12.0-env7kt2iczm4znpvefw4cqb6pbi767a3/include/H5version.h:746:23: error: too few arguments to function 'H5Oget_info3'
   #define H5Oget_info H5Oget_info3
                       ^~~~~~~~~~~~
/tmp/kaig1/spack-stage/spack-stage-petsc-3.7.7-v2q6fsmywzszwwjov2nog6fppzwujepz/spack-src/include/petscviewerhdf5.h:34:37: note: in definition of macro 'PetscStackCallHDF5'
     PetscStackPush(#func);_status = func args;PetscStackPop; if (_status) SETERRQ2(PETSC_COMM_SELF,PETSC_ERR_LIB,"Error in HDF5 call %s() Status %d",#func,(int)_status); \
                                     ^~~~
/tmp/kaig1/spack-stage/spack-stage-petsc-3.7.7-v2q6fsmywzszwwjov2nog6fppzwujepz/spack-src/src/sys/classes/viewer/impls/hdf5/hdf5v.c:729:24: note: in expansion of macro 'H5Oget_info'
     PetscStackCallHDF5(H5Oget_info,(obj, &info));
                        ^~~~~~~~~~~
```

This happens because of an incompatible API change in HDF5 1.12. It affects other packages, too, e.g., #15252.

I can see various options to work around this, ie., limiting hdf5 to :1.10.99, adding `-DH5_USE_110_API`, or updating the code to work with the new API, which in particular for old versions of petsc doesn't seem desirable.

@balay

### PR DESCRIPTION
**Pull request by germasch**
Fixes #15553 

Besides updating release 3.12.4, this fixes a build error by constraining hdf5 to up 1.10 for existing releases.

`develop` and the next release (3.13) will (hopefully) support HDF5 1.12 directly,
so their behavior is unchanged, but for exisiting releases this one makes them build again by
 falling back to HDF5 1.10.

### CODE REVIEW

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 1]
```diff
@@ -145,7 +146,10 @@ class Petsc(Package):
     depends_on('metis@5:~int64', when='@3.8:+metis~int64')
     depends_on('metis@5:+int64', when='@3.8:+metis+int64')
 
-    depends_on('hdf5+mpi+hl+fortran', when='+hdf5+mpi')
+    depends_on('hdf5@:1.10.99+mpi+hl+fortran', when='@:3.12.99+hdf5+mpi')
+    depends_on('hdf5@:1.10.99+mpi+hl+fortran', when='@xsdk-0.2.0+hdf5+mpi')
+    depends_on('hdf5+mpi+hl+fortran', when='@3.13:+hdf5+mpi')
+    depends_on('hdf5+mpi+hl+fortran', when='@develop:+hdf5+mpi')
```
Comment (balay) [inline var/spack/repos/builtin/packages/petsc/package.py]: The second and 4th lines are not needed as spack sorts and handles those versions correctly

Wrt the 3rd line - its probably best to add it after the fix lands in petsc/master branch

Comment (germasch) [reply var/spack/repos/builtin/packages/petsc/package.py]: So what would you like? Just constrain unconditionally?
```
depends_on('hdf5@:1.10.99+mpi+hl+fortran', when='+hdf5+mpi')
```
And you'll fix it up once support is added to your master branch next release? That's just fine with me, I was just trying to address @jedbrown's comment in #15553 by being preemptive about upcoming changes.

Comment (balay) [reply var/spack/repos/builtin/packages/petsc/package.py]: yes will update to hdf5-1.12 once petsc/master is updated.

petsc/master will continue to work with hdf5-1.10 after the fix anyway [so it won't be broken]

Comment (germasch) [reply var/spack/repos/builtin/packages/petsc/package.py]: Okay, force pushed with the simple solution.

**Pull request merged**